core/src/Kokkos_MemoryPool.hpp

```diff
@@ -258,10 +258,56 @@ class MemoryPool {
 
   //--------------------------------------------------------------------------
 
-  KOKKOS_INLINE_FUNCTION_DEFAULTED MemoryPool( MemoryPool && ) = default ;
-  KOKKOS_INLINE_FUNCTION_DEFAULTED MemoryPool( const MemoryPool & ) = default ;
-  KOKKOS_INLINE_FUNCTION_DEFAULTED MemoryPool & operator = ( MemoryPool && ) = default ;
-  KOKKOS_INLINE_FUNCTION_DEFAULTED MemoryPool & operator = ( const MemoryPool & ) = default ;
+  KOKKOS_INLINE_FUNCTION MemoryPool( MemoryPool && rhs )
+    : m_tracker(std::move(rhs.m_tracker))
+    , m_sb_state_array(std::move(rhs.m_sb_state_array))
+    , m_sb_state_size(std::move(rhs.m_sb_state_size))
+    , m_sb_size_lg2(std::move(rhs.m_sb_size_lg2))
+    , m_max_block_size_lg2(std::move(rhs.m_max_block_size_lg2))
+    , m_min_block_size_lg2(std::move(rhs.m_min_block_size_lg2))
+    , m_sb_count(std::move(rhs.m_sb_count))
+    , m_hint_offset(std::move(rhs.m_hint_offset))
+    , m_data_offset(std::move(rhs.m_data_offset))
+  {
+  }
+  KOKKOS_INLINE_FUNCTION MemoryPool( const MemoryPool & rhs )
+    : m_tracker(rhs.m_tracker)
+    , m_sb_state_array(rhs.m_sb_state_array)
+    , m_sb_state_size(rhs.m_sb_state_size)
+    , m_sb_size_lg2(rhs.m_sb_size_lg2)
+    , m_max_block_size_lg2(rhs.m_max_block_size_lg2)
+    , m_min_block_size_lg2(rhs.m_min_block_size_lg2)
+    , m_sb_count(rhs.m_sb_count)
+    , m_hint_offset(rhs.m_hint_offset)
+    , m_data_offset(rhs.m_data_offset)
+  {
+  }
+  KOKKOS_INLINE_FUNCTION MemoryPool & operator = ( MemoryPool && rhs )
+  {
+    m_tracker = std::move(rhs.m_tracker);
+    m_sb_state_array = std::move(rhs.m_sb_state_array);
+    m_sb_state_size = std::move(rhs.m_sb_state_size);
+    m_sb_size_lg2 = std::move(rhs.m_sb_size_lg2);
+    m_max_block_size_lg2 = std::move(rhs.m_max_block_size_lg2);
+    m_min_block_size_lg2 = std::move(rhs.m_min_block_size_lg2);
+    m_sb_count = std::move(rhs.m_sb_count);
+    m_hint_offset = std::move(rhs.m_hint_offset);
+    m_data_offset = std::move(rhs.m_data_offset);
+    return *this;
+  }
+  KOKKOS_INLINE_FUNCTION MemoryPool & operator = ( const MemoryPool & rhs )
+  {
+    m_tracker = rhs.m_tracker;
+    m_sb_state_array = rhs.m_sb_state_array;
+    m_sb_state_size = rhs.m_sb_state_size;
+    m_sb_size_lg2 = rhs.m_sb_size_lg2;
+    m_max_block_size_lg2 = rhs.m_max_block_size_lg2;
+    m_min_block_size_lg2 = rhs.m_min_block_size_lg2;
+    m_sb_count = rhs.m_sb_count;
+    m_hint_offset = rhs.m_hint_offset;
+    m_data_offset = rhs.m_data_offset;
+    return *this;
```

Comment:
```C++
if ( this != other) {
   ...
}
return *this;
```

Reply:
@dalg24 Good point, though I think the self-assignments may be harmless in this case.

Reply:
@dalg24 yes thats a useful optimization in some cases, but I think the current use cases for MemoryPool will never hit this case.